File: spicedb-csharp/examples/SchemaReflection/SchemaReflectionTest.cs
using Xunit;
// Example SchemaReflection demonstrates using schema reflection APIs to
// inspect definitions, compute permissions, find dependent relations, and
// diff schemas.

using SpiceDB.Client;
using static SpiceDB.Client.Consistency;

namespace SchemaReflection;

public class SchemaReflectionTest
{
    private const string Schema = """
        definition user {}

        definition document {
            relation viewer: user
            relation editor: user
            permission view = viewer + editor
            permission edit = editor
        }
        """;

    [Fact]
    public async Task ReflectSchema_ReturnsDefinitions()
    {
        await using var client = SpiceDBClient.CreatePlaintext(SpiceDBTestServer.Endpoint, SpiceDBTestServer.Token);

        await SpiceDBTestServer.ClearDocumentRelationshipsAsync(client);
        await client.WriteSchemaAsync(Schema);

        var result = await client.ReflectSchemaAsync(Full());

        Assert.NotEmpty(result.Revision);

        // The exact names, not just how many. Counting alone passes on a
        // reflection that conflated relations with permissions, or that
        // returned the right number of empty-named entries.
        Assert.Equal(
            new[] { "document", "user" },
            result.Definitions.Select(d => d.Name).OrderBy(n => n).ToArray());

        var docDef = result.Definitions.FirstOrDefault(d => d.Name == "document");
        Assert.NotNull(docDef);
        Assert.Equal(
            new[] { "editor", "viewer" },
            docDef.Relations.Select(r => r.Name).OrderBy(n => n).ToArray());
        Assert.Equal(
            new[] { "edit", "view" },
            docDef.Permissions.Select(p => p.Name).OrderBy(n => n).ToArray());
    }

    [Fact]
    public async Task ComputablePermissions_ForRelation()
    {
        await using var client = SpiceDBClient.CreatePlaintext(SpiceDBTestServer.Endpoint, SpiceDBTestServer.Token);

        await SpiceDBTestServer.ClearDocumentRelationshipsAsync(client);
        await client.WriteSchemaAsync(Schema);

        var (perms, revision) = await client.ComputablePermissionsAsync(
            Full(), "document", "viewer");

        Assert.NotEmpty(revision);

        // `viewer` appears in `view` and nowhere else in this schema, so the
        // answer is exactly one reference -- and it names the permission rather
        // than the relation it was asked about.
        Assert.Equal(
            new[] { "document#view" },
            perms.Select(p => $"{p.DefinitionName}#{p.RelationName}").ToArray());
        Assert.True(perms[0].IsPermission, "expected document#view to be reported as a permission");
    }

    [Fact]
    public async Task DependentRelations_ForPermission()
    {
        await using var client = SpiceDBClient.CreatePlaintext(SpiceDBTestServer.Endpoint, SpiceDBTestServer.Token);

        await SpiceDBTestServer.ClearDocumentRelationshipsAsync(client);
        await client.WriteSchemaAsync(Schema);

        var (deps, revision) = await client.DependentRelationsAsync(
            Full(), "document", "view");

        Assert.NotEmpty(revision);

        // `view = viewer + editor`, so both relations are dependencies and
        // nothing else is.
        Assert.Equal(
            new[] { "document#editor", "document#viewer" },
            deps.Select(d => $"{d.DefinitionName}#{d.RelationName}").OrderBy(n => n).ToArray());
    }

    [Fact]
    public async Task DiffSchema_DetectsChanges()
    {
        await using var client = SpiceDBClient.CreatePlaintext(SpiceDBTestServer.Endpoint, SpiceDBTestServer.Token);

        await SpiceDBTestServer.ClearDocumentRelationshipsAsync(client);
        await client.WriteSchemaAsync(Schema);

        // Diff against a modified schema that adds a new relation and permission
        var newSchema = """
            definition user {}

            definition document {
                relation viewer: user
                relation editor: user
                relation admin: user
                permission view = viewer + editor + admin
                permission edit = editor + admin
                permission manage = admin
            }
            """;

        var (diffs, revision) = await client.DiffSchemaAsync(Full(), newSchema);

        Assert.NotEmpty(revision);

        // newSchema adds one relation and one permission and changes the
        // expression of the other two, so the diff is a known set. Asserting
        // only that it is non-empty would pass on a diff that reported every
        // kind as "unknown" -- the exact outcome of SchemaDiff.FromProto
        // falling through to its default branch.
        var tuples = diffs
            .Select(d => (d.Kind, d.DefinitionName, d.RelationName, d.PermissionName))
            .ToList();
        Assert.Contains(("relation_added", "document", "admin", ""), tuples);
        Assert.Contains(("permission_added", "document", "", "manage"), tuples);
        Assert.Contains(("permission_expr_changed", "document", "", "view"), tuples);
        Assert.DoesNotContain(diffs, d => d.Kind == "unknown");
    }
}

File: spicedb-java/examples/src/test/java/com/authzed/spicedb/examples/SchemaReflectionTest.java
package com.authzed.spicedb.examples;

import com.authzed.spicedb.SpiceDBClient.ComputablePermissionsResult;
import com.authzed.spicedb.SpiceDBClient.DependentRelationsResult;
import com.authzed.spicedb.SpiceDBClient.DiffSchemaResult;
import com.authzed.spicedb.SpiceDBClient.ReflectSchemaResult;
import com.authzed.spicedb.SpiceDBClient.SchemaDefinition;

import org.junit.jupiter.api.Test;

import static com.authzed.spicedb.Consistency.*;
import static org.assertj.core.api.Assertions.*;

/**
 * Demonstrates schema reflection APIs: inspecting definitions, computing
 * permissions, finding dependent relations, and diffing schemas.
 */
class SchemaReflectionTest extends SpiceDBIntegrationTest {

    @Test
    void reflect_schema_returns_definitions() {
        ReflectSchemaResult result = client.reflectSchema(full());

        assertThat(result.revision()).isNotEmpty();
        assertThat(result.definitions()).hasSizeGreaterThanOrEqualTo(2);

        SchemaDefinition docDef = result.definitions().stream()
            .filter(d -> d.name().equals("document"))
            .findFirst()
            .orElseThrow();

        assertThat(docDef.relations()).extracting("name")
            .containsExactlyInAnyOrder("viewer", "editor", "owner");
        assertThat(docDef.permissions()).extracting("name")
            .containsExactlyInAnyOrder("view", "edit", "delete");
    }

    @Test
    void computable_permissions_for_viewer_relation() {
        ComputablePermissionsResult result =
            client.computablePermissions(full(), "document", "viewer");

        assertThat(result.revision()).isNotEmpty();
        assertThat(result.permissions()).isNotEmpty();
        assertThat(result.permissions())
            .extracting("relationName")
            .contains("view");
    }

    @Test
    void dependent_relations_for_view_permission() {
        DependentRelationsResult result =
            client.dependentRelations(full(), "document", "view");

        assertThat(result.revision()).isNotEmpty();
        assertThat(result.relations()).isNotEmpty();
        assertThat(result.relations())
            .extracting("relationName")
            .contains("viewer", "editor", "owner");
    }

    @Test
    void diff_schema_detects_added_relation_and_permission() {
        String newSchema = """
            definition user {}

            definition document {
                relation viewer: user
                relation editor: user
                relation owner: user
                relation admin: user
                permission view = viewer + editor + owner + admin
                permission edit = editor + owner + admin
                permission delete = owner + admin
                permission manage = admin
            }""";

        DiffSchemaResult result = client.diffSchema(full(), newSchema);

        assertThat(result.revision()).isNotEmpty();
        assertThat(result.diffs()).isNotEmpty();
    }
}
